File: solutions/kth_ancestor/sol.cpp
// Written by: le-michael

#include<bits/stdc++.h>
using namespace std;

const int HEIGHT = 100001;
const int WIDTH = 20;
int tree[HEIGHT][WIDTH];
unordered_map<int, int> lookup;

int search(int curr, int kth) {
    if (curr == 0 || kth == 0) return curr;
    if (tree[curr][0] == 0) return 0;
    int lsb = kth & -kth;
    int ind = lookup[lsb];
    return search(tree[curr][ind], kth^lsb);
}

void insert(int x, int y, int ind) {
    if (ind == WIDTH) return;
    tree[x][ind] = y;
    insert(x, tree[y][ind], ind+1);
}

int main() {
    int T, P, Q, x, y, k, task;
    scanf("%d", &T);
    for (int i=0; i<WIDTH; i++) lookup[1<<i] = i;
    while (T--) {
        scanf("%d", &P);
        for (int i=0; i<HEIGHT; i++) {tree[i][0] = 0;}
        while(P--) {
            scanf("%d%d", &x, &y);
            insert(x, y, 0);
        }
        scanf("%d", &Q);
        while(Q--) {
            scanf("%d", &task);
            switch(task) {
                case 0:
                    scanf("%d%d", &y, &x);
                    insert(x, y, 0);
                    break;
                case 1:
                    scanf("%d", &x);
                    tree[x][0] = 0;
                    break;
                case 2:
                    scanf("%d%d", &x, &k);
                    printf("%d\n", search(x, k));
                    break;
            }
        }
    }     
    return 0;
}

File: solutions/kth_ancestor/sol.py
# Written by: le-michael
WIDTH = 20
HEIGHT = 100001
lookup = {}
for i in range(WIDTH): lookup[1 << i] = i

def search(curr, kth):
    if curr == 0 or kth == 0: return curr
    if tree[curr][0] == 0: return 0
    lsb = kth & -kth
    ind = lookup[lsb]
    return search(tree[curr][ind], kth^lsb)

def insert(x, y, ind):
    if ind == WIDTH: return
    tree[x][ind] = y
    insert(x, tree[y][ind], ind+1)

T = int(raw_input())
tree = [[0 for i in range(WIDTH)] for j in range(HEIGHT)]
for _ in range(T):
    P = int(raw_input())
    for i in range(HEIGHT):
        tree[i][0] = 0
    for p in range(P):
        x, y = map(int, raw_input().split())
        insert(x, y, 0)
    Q = int(raw_input())
    for q in range(Q):
        task = map(int, raw_input().split())
        if task[0] == 0:
            insert(task[2], task[1], 0)
        elif task[0] == 1:
            tree[task[1]][0] = 0
        else:
            print search(task[1], task[2])
